rejex.mjs

```diff
@@ -0,0 +1,172 @@
+    return [true, tail]
+  }
+}
+
+// Concatenate two subexpressions together. Only matches if the two
+// subexpressions match.
+export class ConcatExpr {
+  constructor(lhs, rhs) {
+    this.lhs = lhs
+    this.rhs = rhs
+  }
+
+  // Match against a list of character tokens.
+  //
+  // The return value is always a two-element list consisting of
+  // [is_match, remaining_tokens]. The first element is always a bool, and the
+  // second is always a list of tokens.
+  // If there is a match (is_match is true), the matching tokens will be
+  // removed from the remaining_tokens list. Otherwise, the original tokens is
+  // returned in remaining_tokens.
+  match(tokens) {
+    // Try matching both the left and right subexpressions. Do it in order of
+    // left first, then right, so we can allow the left expression to consume
+    // tokens before we pass the remainder into the right.
+    const [lhs_is_match, lhs_remainder] = this.lhs.match(tokens)
+    const [rhs_is_match, rhs_remainder] = this.rhs.match(lhs_remainder)
+
+    // If either subexpressions didn't match, this expression does not match.
+    if (!lhs_is_match || !rhs_is_match) return [false, tokens]
+
+    // This expression is a match. Return the tokens remaining after both the
+    // left and right subexpressions consume them.
+    return [true, rhs_remainder]
+  }
+}
+
+// Alternate two subexpressions together. Matches if either of the
+// subexpressions match.
+export class AlternationExpr {
+  constructor(lhs, rhs) {
+    this.lhs = lhs
+    this.rhs = rhs
+  }
+
+  // Match against a list of character tokens.
+  //
+  // The return value is always a two-element list consisting of
+  // [is_match, remaining_tokens]. The first element is always a bool, and the
+  // second is always a list of tokens.
+  // If there is a match (is_match is true), the matching tokens will be
+  // removed from the remaining_tokens list. Otherwise, the original tokens is
+  // returned in remaining_tokens.
+  match(tokens) {
+    // Try matching both the left and right subexpressions.
+    //
+    // Since there's no ordering to alternation (e.g. [ab] is equivalent to
+    // [ba]), we don't pass the consumed tokens from the one side into the
+    // other.
+    const [lhs_is_match, lhs_remainder] = this.lhs.match(tokens)
+    const [rhs_is_match, rhs_remainder] = this.rhs.match(tokens)
+
+    // If either the left or the right matched, we match
+    if (lhs_is_match || rhs_is_match) {
+      // We can return either the lhs_remainder or the rhs_remainder. We
+      // choose to return the shorter of the two remainders. By doing so,
+      // this expression "consumes" as much of the input as possible, and
+      // therefore is considered "greedy".
+      let shorter_remainder = lhs_remainder
+      if (lhs_remainder.length > rhs_remainder.length) {
+        shorter_remainder = rhs_remainder
+      }
+      return [true, shorter_remainder]
```

Comment:
Maybe mention a quick comparison of using arrays vs objects for the return data types? If it was me, I think I might have used objects {} instead as then the keys help to "document" the code instead of comments, like { is_match: true, remaining_tokens: shorter_remainder }.

Reply:
Switched to using objects.